### DISCUSSION TITLE
Par of Seqs to Seq of Pars Pass: Lab Notebook

### DISCUSSION THREAD

**Discussion by calebmkim** [cucapra/futil#1027] [General]
Lab Notebook for building pass to resolve #840 

Roughly discuss: 
1) What have you been up to this week?
2) What questions, if any, do you have based on that work?
3) What are your plans for next week?

**Comment by calebmkim** [comment]
Writeup for Wed, 6/8:

I worked on a pass to resolve #840 and created a pull request (#1026). Right now it's very conservative about which situations to apply the transformation. 

Questions: (I wrote this on the PR as well). I wasn't sure what the tests should look like? 

Possibly improve the code I have right now to make it more concise. 
Also, I can discuss w/ @rachitnigam about when else to apply the transformation (for example, even if the seq are not all of the same length).

**Comment by calebmkim** [comment]
Fri 6/10: 

I worked on restructuring the code in the way discussed on the PR. 
I also worked on doing the transformation on seqs of different lengths, and adding the @static attribute to the seq that I build. 

Adding these features may complicate my code a bit, so I'm still thinking about how to make it more idiomatic. 

We can discuss what other situations I should look for to apply the transformations for this pass that I can work on next week. I should probably get a clearer plan for what to do next week. 

Edit: After meeting w/ Rachit here a more concrete plan for next week: 
Handle the transformation when only a subset of the Par Block's children fit the criteria for optimization. 
Handle the transformation when seqs are of different lengths, but be more clever about how to form the par blocks (using the "interweaving" we discussed").

**Comment by calebmkim** [comment]
Mon 6/10 

I thought about some ways to find even more situations for the transformation (discussed in #1030). However, it may be smarter to not implement them until we know these situations actually come up in real code. 
I also worked on the issues we discussed: when only a subset of the par block's children fit the criteria for optimization and handling the interweaving. My code is almost done I should have a PR by tomorrow. 

Not too many questions. We can discuss if I should implement the further transformation mentioned in #1030 or not. One idea is to find a heuristic that can detect #1030's situation a good portion of the time but not perfectly.  

Plans: Finish up the code. Also, I will probably need to spend time cleaning the code as well.

**Reply by rachitnigam** [reply]
Sounds awesome! Excited to see the PR

**Comment by calebmkim** [comment]
Wed 6/15 (sorry a day late): 

I submitted a PR for the static-par-conv pass on Tuesday. I then started to think about the next pass (generalizing sharing). I'm going to start a new lab notebook to discuss what I've been thinking.
